src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -1137,11 +1188,89 @@ protected virtual void ReportUnassigned(Symbol symbol, SyntaxNode node, int slot
+            {
+                if (doesNotNeedImplicitFieldInitialization())
+                {
+                    return (error, isSuppressed: false);
+                }
+
+                foreach (var field in _emptyStructTypeCache.GetStructInstanceFields(thisParameter.ContainingType))
+                {
+                    if (_emptyStructTypeCache.IsEmptyStructType(field.Type))
+                        continue;
+
+                    if (field is TupleErrorFieldSymbol)
+                        continue;
+
+                    int slot = VariableSlot(field, thisSlot);
+                    if (slot == -1 || !State.IsAssigned(slot))
+                    {
+                        Debug.Assert((object)field.ContainingType == CurrentSymbol.EnclosingThisSymbol().Type);
+                        AddImplicitlyInitializedField(field);
+                    }
+                }
+
+                if (!compilation.IsFeatureEnabled(MessageID.IDS_FeatureImplicitInitializationInStructConstructors))
+                {
+                    return (error, isSuppressed: false);
+                }
+
+                return (GetEquivalentWarning(error), isSuppressed: true);
+            }
+
+            (ErrorCode, bool isSuppressed) adjustErrorForStructFields(ErrorCode error, Symbol possibleStructField, int fieldSlot)
+            {
+                if (doesNotNeedImplicitFieldInitialization())
+                {
+                    return (error, isSuppressed: false);
+                }
+
+                var thisSlot = GetOrCreateSlot(CurrentSymbol.EnclosingThisSymbol(), createIfMissing: false);
+                var localSlot = fieldSlot;
+                while (localSlot != -1 && RootSlot(localSlot) is int root && root != thisSlot)
+                {
+                    if (root == localSlot)
+                    {
+                        // the offending field access is not contained in 'this'.
+                        return (error, isSuppressed: false);
+                    }
+                    localSlot = root;
+                }
+
+                // the field was contained in 'this' by some level of nesting.
+                // we now have a slot for the field directly contained in 'this'.
+                // should we handle nested fields here? https://github.com/dotnet/roslyn/issues/59890
+                if (localSlot != -1)
+                {
+                    var fieldToInitialize = (FieldSymbol)variableBySlot[localSlot].Symbol;
+                    // We expect this condition to only be false in error scenarios,
+                    // but the erroneous bound node may be further up in the bound tree,
+                    // and not assert-able from here.
+                    // See `FlowDiagnosticTests.AutoPropInitialization5`.
+                    if ((object)fieldToInitialize.ContainingType == CurrentSymbol.EnclosingThisSymbol()!.Type)
```

Comment:
>if ((object)fieldToInitialize.ContainingType == CurrentSymbol.EnclosingThisSymbol()!.Type) [](http://example.com/codeflow?start=20&length=90)

If this check fails, shouldn't we return `(error, isSuppressed: false)`?

Reply:
In this code path, the slot map is telling us that *localSlot is directly contained in 'thisSlot'*. For this reason I feel OK with potentially giving only a warning diagnostic here. Perhaps this could be refactored to be more clear. Also, if the 'localSlot' claims to be contained in 'thisSlot', it seems reasonable to assert that either the related `boundNode.HasErrors` or the `fieldToInitialize.ContainingType` is the same as the `this` symbol's type.

Reply:
I found the check wasn't necessary when we changed 'RootSlot()' to 'variableBySlot[].ContainingSlot'.